src/pushsource/_impl/backend/registry_source.py

```diff
@@ -0,0 +1,171 @@
+    MT_S2_V2,
+    MT_S2_V1,
+    MT_S2_V1_SIGNED,
+    MT_S2_LIST,
+)
+
+from ..helpers import list_argument
+
+LOG = logging.getLogger("pushsource")
+
+
+class RegistrySource(Source):
+    """Uses URIs of container images as source for push items."""
+
+    def __init__(
+        self,
+        dest=None,
+        registry_uris=None,
+        signing_key=None,
+        docker_url="unix://var/run/docker.sock",
+        executor_container_image=None,
+        docker_timeout=None,
+        docker_verify_tls=None,
+        docker_cert_path=None,
+    ):
+        """Create a new source.
+
+        Parameters:
+            dest (str, list[str])
+                The destination(s) to fill in for push items created by this
+                source. If omitted, all push items have empty destinations.
+
+            registry_uris (list[str])
+                List of references to container images with tags+dest tags
+                Format <host>/<namespace>/<repo>:<tag>:<destination_tag>:<destination_tag>
+                Example: registry.redhat.io/ubi:8:latest:8:8.1
+
+            signing_key (list[str])
+                GPG signing key ID(s). If provided, content must be signed
+                using one of the provided keys. Include ``None`` if unsigned
+                should also be permitted.
+                Keys should be listed in the order of preference.
+
+            executor_container_image (str):
+                Path to the container image in which to execute the commands. Must be downloadable
+                without extra permissions.
+
+            docker_url (str):
+                URL of the docker client that should run the container. Local socket by default.
+
+            docker_timeout (int):
+                Timeout for executing Docker commands. Disabled by default.
+
+            docker_verify_tls (bool):
+                Whether to perform TLS verification with the Docker client. Disabled by default.
+
+            docker_cert_path (str):
+                Path to Docker certificates for TLS authentication. '~/.docker' by default.
+        """
+        self._registry_uris = registry_uris
+        self._dest = list_argument(dest)
+        self.signing_keys = list_argument(signing_key)
+        self.executor_container_image = executor_container_image
+        self.docker_url = docker_url
+        self.docker_timeout = docker_timeout
+        self.docker_verify_tls = docker_verify_tls
+        self.docker_cert_path = docker_cert_path
+        self.inspected = {}
+        self.manifests = {}
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        pass
+
+    def _push_item_from_registry_uri(self, uri, signing_key):
+        schema, rest = uri.split("://")
+        host, rest = rest.split("/", 1)
+        repo, tags_part = rest.split(":", 1)
+        tags = tags_part.split(":")
+        source_tag = tags[0]
+        dest_tags = tags[1:]
+        if not dest_tags:
```

Comment:
There is another comment above about "dest", but putting that aside for the moment and just focusing on this one...

I don't think this should be mandatory.  That prevents the caller from making decisions about the destination based on the metadata returned.  If it's mandatory, the caller has to decide up front which destination tag they want on the image when they don't yet know anything about the image except for the URL.  It means the caller of pushsource can't implement some basic things like "use a different tag for SourceContainerImagePushItem vs ContainerImagePushItem".

More generally, it's useful to be able to use these backends without *any* destination involved, e.g. koji source allows this.

Reply:
I don't understand why would user want to have different destination tags for source container vs normal container. User has to know (if that matters to her/him) if image is source container or not because you specify src tags on the input. So one repository can contain both source containers and normal containers but just tagged with different tags.
I don't understand the use case when user wants to decide mapping of src to dst tags after initialization of the source

Reply:
That was an example meant to illustrate that clients might need to set the destination based on info not available up-front, for future use-cases that you and I currently aren't aware of.  It's unnecessarily restrictive, and inconsistent with other sources, to require the caller to specify a destination before you'll let them get any info about the images.

If you want a real use-case which exists today: As a developer using this library, I want to be able to see realistic examples of what this source will return for an image without actually pushing anything so that I can better understand the data produced by this library, and I don't want to be forced to provide some "dummy" destination in order to do that.  That's what I meant by "it's useful to be able to use them without a destination".